src/VisualStudio/TestUtilities2/ProjectSystemShim/Framework/TestEnvironment.vb

```diff
@@ -272,8 +275,7 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.ProjectSystemShim.Fr
             ' Ensure we've pushed everything to the file change watcher
             Dim fileChangeProvider = ExportProvider.GetExportedValue(Of FileChangeWatcherProvider)
             Dim mockFileChangeService = DirectCast(ServiceProvider.GetService(GetType(SVsFileChangeEx)), MockVsFileChangeEx)
-            fileChangeProvider.TrySetFileChangeService_TestOnly(mockFileChangeService)
-            fileChangeProvider.Watcher.WaitForQueue_TestOnly()
+            Task.Run(Function() fileChangeProvider.Watcher.GetTestAccessor().WaitForQueueAsync()).JoinUsingDispatcher(CancellationToken.None)
```

Comment:
Why not just fileChangeProvider.Watcher.GetTestAccessor().WaitForQueueAsync().Wait() here? If that deadlocks because the test thread was unavailable, I'd be pretty concerned that's a functional bug this may mask.

Reply:
This code is executing on the main thread. We can't block it because operations in the queue need it.

Reply:
Exactly, if this is test only, I'm confused what is touching the UI thread at all in the test. If the test implementation isn't safe, something is up.

Reply:
The code always required the UI thread. Previously the test bypassed the UI thread requirement by calling `TrySetFileChangeService_TestOnly` to explicitly give a value to an operation when the UI thread was blocked. The test now uses production code paths to obtain the value, so we account for it here by ensuring the main thread is not blocked.

Reply:
The natural alternative is making this method `async`, which I'll try to do now.

Reply:
This is now rewritten to be asynchronous.